### PR TITLE
[5G: MVC] [SessionD] Support for UE initiated Session Release

### PR DESCRIPTION
**Pull request by sanjay-ACL**
Signed-off-by: sanjay-ACL <sanjaya.k@altencalsoftlabs.com>

## Summary
This code to handle release message from UE through AMF.

## Test Plan

[Establishment_Release_logs.txt](https://github.com/magma/magma/files/5319036/Establishment_Release_logs.txt)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 1]
```diff
@@ -193,8 +193,11 @@ uint32_t SessionState::get_current_version() {
   return current_version_;
 }
 
-void SessionState::set_current_version(int new_session_version) {
+void SessionState::set_current_version(int new_session_version,
+	           SessionStateUpdateCriteria& uc) {
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionState.cpp]: `uc` -> `session_uc` please. :) We are trying to be more consistent with naming parameters.

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 2]
```diff
@@ -220,7 +229,25 @@ std::vector<SetGroupFAR>& SessionState::get_all_far_rules() {
 void SessionState::sess_infocopy(struct SessionInfo* info) {
   // Static SessionInfo vlaue till UPF node value implementation
   // gets stablized.
-  info->state               = magma::lte::Fsm_state_FsmState_CREATING;
+  int curr_state = get_state();
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionState.cpp:232]: I would define a separate function like `fsm_state_to_proto`. But why don't we simply use the FSM state proto value directly? Then we don't have to do any conversion code. Maybe add that as a TODO if we want to move forward with that.

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 3]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+ * Go over SessionState vector and dfnd the respective dnn (apn) 
+ * start terminating session process
+ */
+bool SessionStateEnforcer::m5g_release_session(
+      SessionMap& session_map, const std::string& imsi,
+      const std::string& dnn, SessionUpdate& session_update)
+{
+   bool exist = false;
+   auto map_element = session_map.find(imsi);
+   if((map_element == session_map.end())) {
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Might be good to look into the `find_session` function in `SessionStore.h`

**Review by themarwhal — CHANGES_REQUESTED** [review]

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 4]
```diff
@@ -115,15 +119,9 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
-  SessionState::SessionInfo sess_info;
-  sess_info.imsi       = imsi;
-  sess_info.ip_addr    = ip_addr;
-  sess_info.Pdr_rules_ = session_state.get_all_pdr_rules();
-  sess_info.Far_rules_ = session_state.get_all_far_rules();
-  session_state.sess_infocopy(&sess_info);
 
   /* session_state elments are filled with rules. State needs to be
-   * moved to CREATED and sending message to UPF.
+   * moved to CREATED, increament version and send message to UPF.
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: increment*

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 5]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+		  << " nothing to release. Return ";
+     return false;
+  }
+  return true;
+}
+
+/*Start processing to terminate respective session requested from AMF*/
+void SessionStateEnforcer::m5g_start_session_termination(
+      const std::string& imsi, const std::unique_ptr<SessionState>& session,
+      const std::string& dnn, SessionStateUpdateCriteria& uc)
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: `uc` -> `session_uc`

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 6]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+{
+   auto map_element = session_map.find(imsi);
+   if(map_element == session_map.end()) {
+      MLOG(MERROR) << "No session found in SessionMap for IMSI " << imsi
+                   << " during terminating session_id " << session_id
+		   << "skipping termination";
+      return;
+   }
+
+   SessionStateUpdateCriteria& uc = session_update[imsi][session_id];
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: `uc` -> `session_uc`

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 7]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp:284]: Again, let's use the `find_session` from SessionStore.h

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 8]
```diff
@@ -206,6 +209,12 @@ void SessionState::insert_far(SetGroupFAR* rule) {
   FarList_.push_back(*rule);
 }
 
+/* Remove all Pdr, FAR rules */ 
+void SessionState::remove_all_rules() {
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionState.cpp:213]: Do we not have to update the updatecriteria for this action?

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 9]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
    */
   auto update_criteria = get_default_update_criteria();
   session_state.set_fsm_state(CREATED, update_criteria);
-  MLOG(MDEBUG) << "State of session changed to "
-               << session_fsm_state_to_str(session_state.get_state());
+  uint32_t cur_version = session_state.get_current_version();
+  session_state.set_current_version(cur_version++, update_criteria);
+  MLOG(MINFO) << "State of session changed to "
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp:139]: Might be useful to include this log message in the set_current_version function

**Comment by themarwhal** [comment]
Please mark the PR as open so it shows up in my PR list :)

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 10]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+
+   for (auto& session : map_element->second) {
+     auto config     = session->get_config();
+     auto session_id = session->get_session_id();
+     if (config.common_context.apn() == dnn){
+	/*Irrespective of any State of Session, release and terminate*/
+        SessionStateUpdateCriteria& uc = session_update[imsi][session_id];
+        MLOG(MINFO) 
+		<< "Start terminating the release request for IMSI "
+                << imsi << " having session id " << session_id
```
Comment (uri200) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Use only session_id. We build the sessions id using the imsi (unless you are doing something different for 5g).  SessionId will be enough to uniquely identify the UE and the seessions and we save some characters on that log

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 11]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+		<< " current state of session is "
```
Comment (uri200) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: can we make this log smaller?
```
       MLOG(MINFO) << 
        "Trying to release session " << session_id  << 
		" from state "  << session_fsm_state_to_str(session->get_state());
```

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 12]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+                << session_fsm_state_to_str(session->get_state());
+	exist = true;
```
Comment (uri200) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: as recommended by @themarwhal if you use `find_session` she created. It is very useful and you will save lots of code and variables like this

```
    SessionSearchCriteria criteria(imsi, IMSI_AND_APN, session_id);
    auto session_it = find_session(session_map, criteria);
    if (!session_it) {
      MLOG(MERROR)
          << session_id
          << " not found when m5g_release_session";
      return False;
    }

	auto& session = **session_it;
	auto session_id = session->get_session_id()

```

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 13]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+   * current_state_ and
+   * dnn can directly be used.
+   */
+  m5g_remove_rules_from_session_for_termination(imsi,
+		   session, dnn, uc);
+
+  /* Forcefully terminate session context on time out
+   * time out = 5000ms from sessiond.yml config file
+   */
+  MLOG(MINFO) << "Scheduling a force termination timeout for session_id "
```
Comment (uri200) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp:218]: lets set this as `MDEBUG` as we do in 4g

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 14]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+ */
+void SessionStateEnforcer::m5g_update_session_state_message_from_upf(
+		const std::string& imsi, const std::string& session_id,
+		SessionFsmState new_state)
+{
+  auto session_map    = session_store_.read_sessions({imsi});
+  auto session_update = 
+	     SessionStore::get_default_session_update(session_map);
+  auto map_element = session_map.find(imsi);
+  if(!(map_element == session_map.end())) {
```
Comment (uri200) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: If session is not found we don't log anything. Maybe it is better to do 
```
if(map_element == session_map.end()){
	MLOG(MWARNING) << sessionid << " session not found while m5g_update_session_state_message_from_upf " 
	return;
}

```

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp [hunk 15]
```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
...
+        MLOG(MINFO) << "Based on UPF message state of session changed to "
+               << session_fsm_state_to_str(new_state)
+	       << " current version is " << cur_version;
+        bool update_success = session_store_.update_sessions(session_update);
+        if(update_success) {
+          MLOG(MINFO) << "Updated SessionStore SessionState based on UPF message " 
+	              << "of IMSI " << imsi << " session_id" << session_id;
+        } else {
+          MLOG(MERROR) << "Failed to update SessionStore based on UPF message"
+	               << "of IMSI " << imsi << " session_id" << session_id;
```
Comment (uri200) [inline lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: IMSIS are redundant and will make logs longers. I recommend to remove any imsi and use session_id only

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 2 above)
Comment (gurrapualt) [reply lte/gateway/c/session_manager/SessionState.cpp:232]: Noted comments, looks resonable.

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 8 above)
Comment (gurrapualt) [reply lte/gateway/c/session_manager/SessionState.cpp:213]: As we are doing .. in starting probably not required, later once we integrate with  policydb script then we might needed.

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 1 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionState.cpp]: Adapted

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 4 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: changed

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 5 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Adapted

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 15 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: +1 ^

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 6 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Done

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 9 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp:139]: Good point to make it generic, but we would like to know, by which procedure the version gets incremented. In my point of view, lets print where ever required.

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 10 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Adapted.

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 11 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Done

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 12 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Adapted with recommended function on possible places

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 14 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Adapted with little change
 MLOG(MWARNING) << "Session id (" << session_id << ") not found to update"
                   << " SessionStore based on UPF message";

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 15 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Adapted

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 13 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp:218]: Currently we are using MINFO to get the messages during testing, else we need to change the log level in config file every time to test. We will do the needful of changing the log level once the basic code gets stabilized.

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 3 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp]: Yes. Adapted in function m5g_release_session  along with in function m5g_update_session_state_message_from_upf

File: lte/gateway/c/session_manager/SessionStateEnforcer.cpp (hunk 7 above)
Comment (sanjay-ACL) [reply lte/gateway/c/session_manager/SessionStateEnforcer.cpp:284]: find_session returns vector of elements of SessionState. The function m5g_complete_termination, need to work on SessionMap as well, to remove entry of imsi from map. Hence I am thinking not to be adapted as code will be repeated. 
Nevertheless find_session had been adapted in other 2 functions m5g_release_session and m5g_update_session_state_message_from_upf.

**Review by uri200 — APPROVED** [review]
LGTM

**Review by themarwhal — APPROVED** [review]

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 2 above)
Comment (gurrapualt) [reply lte/gateway/c/session_manager/SessionState.cpp:232]: Sure, I am planning to do this change as part of another PR, that I'll raise.

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 8 above)
Comment (gurrapualt) [reply lte/gateway/c/session_manager/SessionState.cpp:213]: Not in this case, as its called before we move to "created" state.  Not required, update creteria here.

**Comment by themarwhal** [comment]
Hi @sanjay-ACL and @gurrapualt can we fix the failure on Jenkins? When it is resolved, I can go ahead and merge.

**Comment by sanjay-ACL** [comment]
> Hi @sanjay-ACL and @gurrapualt can we fix the failure on Jenkins? When it is resolved, I can go ahead and merge.

Hi @themarwhal Marie, I think, we need to re trigger the Jenkin build with latest pulled code by doing false pushing code. Amy be a extra comment line we need to add and push.  This we will do today.

**Comment by sanjay-ACL** [comment]
> Hi @sanjay-ACL and @gurrapualt can we fix the failure on Jenkins? When it is resolved, I can go ahead and merge.

 @themarwhal , Hi Marie, Looks like "Jenkins — Tests successfully passed." Pls. proceed and merge to master.

**Comment by themarwhal** [comment]
Please fix the SessionD build failure seen in both LTE/CWF Integration test logs. @sanjay-ACL

**Comment by sanjay-ACL** [comment]
> Please fix the SessionD build failure seen in both LTE/CWF Integration test logs. @sanjay-ACL

@themarwhal , I see the pipelined issue, will request @gurrapualt  (Venu) to look at this and fix the same.

**Comment by gurrapualt** [comment]
This code pulled as part of https://github.com/magma/magma/pull/3366   request, so closing this.

**Pull request closed without merging**